### PR TITLE
Timer class will call setTimeout with a negative integer depending on race outcome

### PR DESCRIPTION
**Pull request by ParmesanoRaygiono**
According to the W3C spec, it actually looks like the standard says that a negative timeout into setTimeout should be treated as a 0ms delay timer, so strictly speaking, there is nothing wrong with Fluid scheduling a job with a negative timeout. It is well-defined behavior.

Still, I am sharing the condition of Fluid's Timer class that ends up causing setTImeout to be called in case this is something we are interested in addressing, since scheduling a job with a negative timeout is conceptually strange.

The condition is this:
Code schedules a job with a Fluid Timer for X seconds. Before X seconds has elapsed (let's call the amount of seconds that has elapsed Y, such that Y < X), someone calls .restart() on the Timer, causing the timer's restart info field to be filled.

Now, if the first time the JSEngine flushes the queued tasks is after X + Y seconds have elapsed (let's call the total amount of time elapsed Z, timer.handler() will decide that "yes, we were asked to restart, let's schedule another task with setTimeout with a duration of X +Y - Z". Notice that this is a negative number in this case.

In plain English, it doesn't look like Timer's restart logic is explicitly watching out for the case where the first time the scheduled handler gets to execute occurs after ultimately the desired point in time (taking into account one or more restarts)

The proposed fix is to watch out for `Timer.handler()` inadvertently triggering a `setTimeout` with a negative duration. Instead, if the timer realizes that it would have asked for a negative duration when attempting to fulfill the restart, it should just execute the requested lambda immediately. 

Explicitly setting "allow edits by maintainers" to true because I would love some assistance on validating this idea.

### CODE REVIEW

File: common/lib/common-utils/src/timer.ts [hunk 1]
```diff
@@ -200,13 +200,20 @@ export class Timer implements ITimer {
...
             const remainingTime = this.calculateRemainingTime(restart);
-            this.startCore(remainingTime, () => restart.handler(), restart.duration);
-        } else {
-            // Run clear first, in case the handler decides to start again
-            const handler = this.runningState.handler;
-            this.clear();
-            handler();
+
+            if (remainingTime > 0)
+            {
```
Comment (tyler-cai-microsoft) [inline common/lib/common-utils/src/timer.ts:205]: My bet is that the linter is complaining that this bracket is on the next line and that's why the ci is failing. @tylerbutler it's not immediately obvious the next step - `npm run prettier && npm run eslint` fails, what command should be run to fix it?

Comment (tyler-cai-microsoft) [reply common/lib/common-utils/src/timer.ts:205]: Looking at the `package.json` file I would assume you would run `npm run format`

Comment (Abe27342) [reply common/lib/common-utils/src/timer.ts:205]: `npm run prettier:fix` should do it

Comment (Abe27342) [reply common/lib/common-utils/src/timer.ts:205]: (`lint:fix` script should work too; that way you only need to remember one thing)

**Comment by tyler-cai-microsoft** [comment]
OK, for tests, in the `common/lib/common-utils/src/test/mocha/timer.spec.ts` file a test to add under the `describe("Timer"` tests would look something like
```
        it("Should immediately execute at negative time", () => {
            const overrideTimeout = -100; // a negative time
            const initialRunCount = runCount;
            timer.start(overrideTimeout);
            assert.strictEqual(runCount, initialRunCount + 1, "Should have executed immediately");
        });

**Comment by ParmesanoRaygiono** [comment]
> OK, for tests, in the `common/lib/common-utils/src/test/mocha/timer.spec.ts` file a test to add under the `describe("Timer"` tests would look something like
> 
> ```
>         it("Should immediately execute at negative time", () => {
>             const overrideTimeout = -100; // a negative time
>             const initialRunCount = runCount;
>             timer.start(overrideTimeout);
>             assert.strictEqual(runCount, initialRunCount + 1, "Should have executed immediately");
>         });
> ```

Thanks! I can certainly add this test.

Because the new code is under `if (restart !== undefined)`, should we have a test that covers having a negative duration after a restart has been queued internally? Or does your test accomplish that?

**Comment by tyler-cai-microsoft** [comment]
> > OK, for tests, in the `common/lib/common-utils/src/test/mocha/timer.spec.ts` file a test to add under the `describe("Timer"` tests would look something like
> > ```
> >         it("Should immediately execute at negative time", () => {
> >             const overrideTimeout = -100; // a negative time
> >             const initialRunCount = runCount;
> >             timer.start(overrideTimeout);
> >             assert.strictEqual(runCount, initialRunCount + 1, "Should have executed immediately");
> >         });
> > ```
> 
> Thanks! I can certainly add this test.
> 
> Because the new code is under `if (restart !== undefined)`, should we have a test that covers having a negative duration after a restart has been queued internally? Or does your test accomplish that?

Good point, we would need another test. Could you do the same code above, except with `timer.restart` instead of `timer.start`?

**Comment by tyler-cai-microsoft** [comment]
> > 
> 
> Caught up offline, here is what we talked about: I think it's a little bit more complicated than that because we need to simulate the handler being late the first time it gets to execute. See Commit 3 for my attempt at testing it in the test-suite

@curtisman would you know anything about this?

**Comment by curtisman** [comment]
@tyler-cai-microsoft I am not sure what the question is.  Are you asking how to write a test that could exercise this?

**Comment by tyler-cai-microsoft** [comment]
> @tyler-cai-microsoft I am not sure what the question is. Are you asking how to write a test that could exercise this?

Yeah, that's what I'm asking

**Comment by ParmesanoRaygiono** [comment]
@curtisman I think the latest commits has the test that would execute the code I added by deliberately delaying the flushing of the promises until the clock has moved way past the original scheduled time, accounting for the restart.

Since passing in a negative value into `setTimeout` is standard conforming, in a conforming JSEngine, this new test should in theory pass with or without my product code change. Is it worth polyfilling `setTimeout` for this particular test to validate that a negative value wasn't passed in?
